browser/lib/transport/xhrrequest.js

```diff
@@ -211,7 +211,7 @@ var XHRRequest = (function() {
 				return;
 			}
 
-			var err = responseBody.error;
+			var err = responseBody.error && ErrorInfo.fromValues(responseBody.error);
```

Comment:
This confuses me. Looks like a boolean but later on is assigned to a new `ErrorInfo`. What's going on here? Might some commentary help in the code to explain? Or feel free to tell me I'm just being thick! 😁

Reply:
the standard error response body from Ably looks like
```
{
	"error": {
		"message": "Could not find path: /nonexistent. (See https://help.ably.io/error/40400 for help.)",
		"code": 40400,
		"statusCode": 404,
		"href": "https://help.ably.io/error/40400",
		"serverId": "frontend.912a.2.eu-west-1-A.i-0a525ab4287758c86.108ykbPEgAoyLO"
	}
}
```
so this is just presenting that as an ErrorInfo rather than a plain object, which makes it display a bit nicer.

This format is documented in https://www.ably.io/documentation/rest-api#general, but not really in the spec anywhere that I can see. (kindof sortof [TI2](https://docs.ably.io/client-lib-development-guide/features/#TI2), but that's way too vague/handwavey to actually be useful)

Reply:
Thanks, @SimonWoolf ... so just to make sure I understand right, `responseBody.error` is evaluated as a boolean (`null` == `false`) but the statement as a whole returns an object?

Reply:
yes, exactly

Reply:
@QuintinWillison these are truthy and falsy values. But, we also need to consider edge cases while using them. A small read will be helpful https://www.sitepoint.com/javascript-truthy-falsy/

Reply:
The library is assuming, as it's entitled to, that the server is returning response bodies in the form that it expects per [our public api](https://www.ably.io/documentation/rest-api#general), which defines that the `error` field if present will be an object with at least the core three fields. JS Objects are always truthy.